### PR TITLE
feat: implement Headers class

### LINKED ISSUE #12: Implement Headers
Implement & test [Headers](https://fetch.spec.whatwg.org/#headers-class) from the fetch spec.

Probably a new file in `src` called `Headers.ts` with a matching `Headers.test.ts` file. It should implement the `Headers` class from the `whatwg-fetch` typings:

```js
class Headers implements Headers {
   // ...
}
```

### PR DESCRIPTION
**Pull request by ahume**
* Not yet Iterable.
* Constructor should support passing Headers or sequence<sequence<ByteString>> or OpenEndedDictionary<ByteString>.

Fix #12.

### CODE REVIEW

File: src/Headers.ts [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+export class Headers implements Headers {
+  private headers: HeadersList = [];
+  private guard = Guards.none;
+
+  // HTTP
+  // HTTP whitespace bytes "0x09, 0x0A, 0x0D, and 0x20"
+
+  constructor(headers?: any) {
+    // TODO: Support List and Headers passed in.
+    for (let name in headers) {
```
Comment (ahume) [inline src/Headers.ts]: iterate safely.

File: src/Headers.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    if (this.guard === Guards.response &&
+      this.isForbiddenResponseHeaderName(name)) {
+        return;
+    }
+
+    var set = false;
+
+    for (var i = 0; i < this.headers.length; ++i) {
```
Comment (ahume) [inline src/Headers.ts]: Comment.

**Comment by ahume** [comment]
Class has to be iterable to allow 
```
new Headers(new Headers({ "h1":"v1"}));

**Comment by phuu** [comment]
Discussed the guard issue with Anne in whatwg/fetch#182. I think we can do it by creating some Headers constructor factories that close over particular guard types, rather than introducing a public API.

Otherwise this is looking really nice!

File: src/Headers.test.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
+/// <reference path="./.d.test.ts" />
+"use strict";
+import test = require("ava");
+import { Headers, ImmutableHeaders, RequestHeaders, RequestNoCorsHeaders, ResponseHeaders } from "./Headers";
```
Comment (phuu) [inline src/Headers.test.ts]: This line's too long. The style I've seen before is:

```js
import {
  X,
  Y,
  Z
} from "./blap";
```

File: src/Headers.test.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+test("is requireable", (t: TestAssertions) => {
```
Comment (phuu) [inline src/Headers.test.ts]: "Headers is requireable"

File: src/Headers.test.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+test("append() fails if guard is request-no-cors and header is not simple", (t: TestAssertions) => {
+  let headers = new RequestNoCorsHeaders();
+  headers.append("Header-1", "Value-1");
+  headers.append("Content-Language", "Value-2");
+  t.false(headers.has("Header-1"));
+  t.same(headers.get("Content-Language"), "Value-2");
+  t.end();
+});
+
+test("append() succeeds if guard is request-no-cors and header name is simple with allowed value", (t: TestAssertions) => {
```
Comment (phuu) [inline src/Headers.test.ts]: Line too long (there are a few others).

```js
test(
  "boom bap",
  (t) => {
    // ...
  }
);
```

File: src/Headers.ts [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+"use strict";
+
+export interface Header {
+  [name: string] : string;
+}
+
+export interface HeadersList extends Array<Header> {};
+
+// String based enum
+const Guards = {
```
Comment (phuu) [inline src/Headers.ts]: Needed?

File: src/Headers.ts [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+      }
+      for (let name in headers) {
+        if (headers.hasOwnProperty(name)) {
+          this.append(name, headers[name]);
+        }
+      }
+    }
+  }
+
+  setGuard(guard: string): void {
```
Comment (phuu) [inline src/Headers.ts]: Not needed.

File: src/Headers.ts [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+
+export class RequestHeaders extends AbstractHeaders {
+  constructor(headers?: any) {
+    super(headers);
+    this.setGuard(Guards.request);
+  }
+}
+
+export class RequestNoCorsHeaders extends AbstractHeaders {
+  constructor(headers?: any) {
```
Comment (phuu) [inline src/Headers.ts]: Just overwrite the guard property of the AbstractHeaders type.

File: src/Headers.test.ts (hunk 3 above)
Comment (ahume) [inline src/Headers.test.ts]: Done.

File: src/Headers.test.ts (hunk 4 above)
Comment (ahume) [inline src/Headers.test.ts]: Done.

File: src/Headers.test.ts (hunk 5 above)
Comment (ahume) [inline src/Headers.test.ts]: Done.

File: src/Headers.ts (hunk 6 above)
Comment (ahume) [inline src/Headers.ts]: Better than strings all over the place. Could use enum but still need to look up strings, so meh.

File: src/Headers.ts (hunk 7 above)
Comment (ahume) [inline src/Headers.ts]: Done.

File: src/Headers.ts (hunk 8 above)
Comment (ahume) [inline src/Headers.ts]: Done.

File: src/Headers.ts (hunk 6 above)
Comment (ahume) [inline src/Headers.ts]: Changed to enum as never exposed now.

**Pull request merged**